libraries/chain/apply_context.cpp

```diff
@@ -512,29 +519,33 @@ void apply_context::schedule_deferred_transaction( const uint128_t& sender_id, a
 
       // Use remove and create rather than modify because mutating the trx_id field in a modifier is unsafe.
       db.remove( *ptr );
```

Comment:
My assumption, based off of the specification, is that there are two ways to enter deferred transaction logic:

1. Through `transaction_api::send_deferred`.
2. Through a user extending his/her code via a plugin and calling `apply_context::schedule_deferred_transaction`.

Therefore two individual (seemingly redundant) checks must be made. Am I wrong in this assumption?

Reply:
I don't think apply_context::schedule_deferred_transaction should ever be called outside of wasm.  Also, the main point that I was trying to make was that the behavior is not the same in the two cases.

Reply:
I believe they now have the same behavior.

Reply:
The behavior is still different in the case of `replace_existing = true`.  `transaction_api::send_deferred` short-circuits this case to a no-op, but `apply_context::schedule_deferred_transaction` handles it in a more complex way.

Reply:
Yes, that is what I was told to implement.

> [XX] `send_deferred`
> [XX]   - If `replace_existing` == true in `send_deferred`; it shall be a noop

> [XX] `schedule_deferred_transaction`
> [XX]   - If there is no transaction to replace: fail
> [XX]   - If there is a transaction to replace: continue to cancel the old one but do not schedule the new one

Hence, the assertion in `apply_context::schedule_deferred_transaction`.

Reply:
Can you explain the motivation for this difference?

Reply:
The motivation is unbeknownst to me. I just took it at face-value.

Reply:
I'm somewhat concerned that this may be a misunderstanding.  The spec for `schedule_deferred_transaction` sounds to me like reasonable behavior for both of them.  @arhag